### PR TITLE
feat: add hooks for `OverrideInit`

### PR DESCRIPTION
**Pull request by AayushSabharwal**
## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API

### CODE REVIEW

File: lib/NonlinearSolveBase/src/solve.jl [hunk 1]
```diff
@@ -264,11 +314,20 @@ end
 
 function SciMLBase.__init(
         prob::AbstractNonlinearProblem, alg::AbstractNonlinearSolveAlgorithm, args...;
+        initializealg = NonlinearSolveDefaultInit(),
         kwargs...
 )
-    return NonlinearSolveNoInitCache(prob, alg, args, kwargs)
+    cache = NonlinearSolveNoInitCache(
+        prob, alg, args, kwargs, initializealg, ReturnCode.Success)
```
Comment (AayushSabharwal) [inline lib/NonlinearSolveBase/src/solve.jl:321]:
```suggestion
        prob, alg, args, kwargs, initializealg, ReturnCode.Default)
```

Probably should be this

File: lib/NonlinearSolveFirstOrder/src/solve.jl [hunk 2]
```diff
@@ -206,13 +208,17 @@ function SciMLBase.__init(
...
             fu, u, u_cache, prob.p, du, J, alg, prob, globalization,
             jac_cache, descent_cache, linesearch_cache, trustregion_cache,
             stats, 0, maxiters, maxtime, alg.max_shrink_times, timer,
-            0.0, true, termination_cache, trace, ReturnCode.Default, false, kwargs
+            0.0, true, termination_cache, trace, ReturnCode.Default, false, kwargs,
+            initializealg
         )
     end
+
+    NonlinearSolveBase.run_initialization!(cache)
```
Comment (AayushSabharwal) [inline lib/NonlinearSolveFirstOrder/src/solve.jl:220]: Should this be inside the `@static_timeit`?

File: lib/NonlinearSolveQuasiNewton/src/solve.jl [hunk 3]
```diff
@@ -204,15 +215,18 @@ function SciMLBase.__init(
...
             initialization_cache, descent_cache, linesearch_cache,
             trustregion_cache, update_rule_cache, reinit_rule_cache,
             inv_workspace, stats, 0, 0, alg.max_resets, maxiters, maxtime,
             alg.max_shrink_times, 0, timer, 0.0, termination_cache, trace,
-            ReturnCode.Default, false, false, kwargs
+            ReturnCode.Default, false, false, kwargs, initializealg
         )
     end
+
+    NonlinearSolveBase.run_initialization!(cache)
```
Comment (AayushSabharwal) [inline lib/NonlinearSolveQuasiNewton/src/solve.jl:228]: Again with the `@static_timeit`

File: lib/NonlinearSolveSpectralMethods/src/solve.jl [hunk 4]
```diff
@@ -145,13 +148,16 @@ function SciMLBase.__init(
...
+        cache = GeneralizedDFSaneCache(
             fu, fu_cache, u, u_cache, prob.p, du, alg, prob,
             σ_n, T(alg.σ_min), T(alg.σ_max),
             linesearch_cache, stats, 0, maxiters, maxtime, timer, 0.0,
-            tc_cache, trace, ReturnCode.Default, false, kwargs
+            tc_cache, trace, ReturnCode.Default, false, kwargs, initializealg
         )
     end
+
+    NonlinearSolveBase.run_initialization!(cache)
```
Comment (AayushSabharwal) [inline lib/NonlinearSolveSpectralMethods/src/solve.jl:159]: `@static_timeit`

**Comment by avik-pal** [comment]
Can you briefly mention the motivation behind this? I am not too familiar with the background (other PRs for DAEs and stuff)

**Comment by AayushSabharwal** [comment]
Sure. MTK currently generates initialization problems for ODEs. https://github.com/SciML/ModelingToolkit.jl/pull/3253 aims to expand this to all such problems. The basic use case is that users can specify initial conditions for variables that MTK eliminates from the system, and the initialization process solves for the states of the system. As a contrived example, consider:

```julia
y ~ x^3
x^2 + y^2 ~ 2
```
MTK will reduce this to a system with a single unknown `x` and a single equation (the latter). It will dynamically calculate `y` inside the `f` provided to `NonlinearFunction`. Currently, the user _has_ to specify an initial condition for `x`. With `OverrideInit`, the user can specify an initial condition for `y` (say 3) and MTK will generate the following `InitializationProblem` (`NonlinearProblem`):

```julia
y ~ 3
y ~ x^3
```

and solve for `x`. As an added check, it ensures overspecified initial conditions are consistent. e.g. in the example above, if the initial condition `[x => 3, y => 3]` is specified, MTK generates a `NonlinearLeastSquaresProblem`:

```julia
x ~ 3
y ~ 3
y ~ x^3
```

This fails the nonlinear solve with `ReturnCode.InitialFailure` telling the user their initial conditions are inconsistent/unsatisfiable.

MTK can also potentially solve for parameters, such as if `x + y ~ p^2` is an equation, the user can specify `x` and `y` and MTK will generate a `NonlinearProblem` to solve for `p`. 

The core infrastructure is in SciMLBase, but all solver libraries need to hook into it. In most cases, this is a trivial change: https://github.com/SciML/StochasticDelayDiffEq.jl/pull/84/files. NonlinearSolve has a billion entry points, multiple different "integrators" (caches) and a whole host of other intricacies which make this PR more complicated.

File: lib/NonlinearSolveFirstOrder/src/solve.jl (hunk 2 above)
Comment (avik-pal) [reply lib/NonlinearSolveFirstOrder/src/solve.jl:220]: yes

**Review by ChrisRackauckas — APPROVED** [review]

**Review by avik-pal — APPROVED** [review]

**Comment by avik-pal** [comment]
@AayushSabharwal feel free to tag the appropriate releases. All of the (sub)-packages probably need a version bump I think

**Comment by AayushSabharwal** [comment]
I'll PR to bump versions, but I don't have write access to be able to tag them

**Pull request merged**